src/hooks/usePermitAllowance.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  AllowanceData,
+  AllowanceProvider,
+  AllowanceTransfer,
+  MaxAllowanceTransferAmount,
+  PERMIT2_ADDRESS,
+  PermitSingle,
+} from '@uniswap/permit2-sdk'
+import { Token } from '@uniswap/sdk-core'
+import { useWeb3React } from '@web3-react/core'
+import ms from 'ms.macro'
+import { useCallback, useEffect, useMemo, useState } from 'react'
+
+const PERMIT_EXPIRATION = ms`30d`
+const PERMIT_SIG_EXPIRATION = ms`30m`
+
+function toDeadline(expiration: number): number {
+  return Math.floor((Date.now() + expiration) / 1000)
+}
+
+export function usePermitAllowance(token?: Token, spender?: string) {
+  const { account, provider } = useWeb3React()
+  const allowanceProvider = useMemo(() => provider && new AllowanceProvider(provider, PERMIT2_ADDRESS), [provider])
+  const [allowanceData, setAllowanceData] = useState<AllowanceData>()
+
+  useEffect(() => {
+    if (!account || !token || !spender) return
+
+    allowanceProvider
+      ?.getAllowanceData(token.address, account, spender)
+      .then((data) => {
+        if (stale) return
+        setAllowanceData(data)
+      })
+      .catch((e) => {
+        console.warn(`Failed to fetch allowance data: ${e}`)
+      })
+
+    let stale = false
+    return () => {
+      stale = true
+    }
+  }, [account, allowanceProvider, spender, token])
+
+  return allowanceData
+}
+
+interface Permit extends PermitSingle {
+  sigDeadline: number
+}
+
+export interface PermitSignature extends Permit {
+  signature: string
+}
+
+export function useUpdatePermitAllowance(
+  token?: Token,
```

Comment:
can't we just require these values vs making them optional and throwing?

Reply:
this is easier because of hooks